Reals

lsk567 opened the pull request:
This PR adds the `real` datatype to Uclid.

adwait reviewed (commented):
Hey Shaokai, I probably missed this, but did we discuss `r` as an alternative to the `real` suffix (`real` is kind of cumbersome)?

polgreen commented on test/test-real-1.ucl:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  // Part 2: System specification.
+  // fp compare have not been implemented yet
```
fp compare? or real compare?

lsk567 replied:
Thanks for catching this! This stale comment is now deleted.

polgreen reviewed (changes requested):
Cool, thanks. Broadly looks good and like it's following the same patterns as integerlits. 

Couple of bits where Reals still need to be handled (I think fixing these will make the regression test pass):
- I'm a bit confused about where RealLit is defined. I think it needs to be in UclidLexical, so I wonder if you've just forgotten to commit a line? As it is currently written, it throws a compiler warning saying that "fruitless type test: a value of type UclidTokenParsers.this.Elem cannot also be a uclid.lang.RealLit" 

- add reals to this match in the TypeChecker: https://github.com/uclid-org/uclid/blob/reals/src/main/scala/uclid/lang/TypeChecker.scala#L222
- and this: https://github.com/uclid-org/uclid/blob/reals/src/main/scala/uclid/smt/Context.scala#L153
- Also here (actually this one also fails on float, so if you could add float as well that would be fantastic): https://github.com/uclid-org/uclid/blob/reals/src/main/scala/uclid/smt/Z3Interface.scala#L306


Also, some more testing please!
- Can you add some tests that should fail verification as well as tests that pass verification?
- Can you also add the tests to run using the Z3 interface as well as the SMTlib interface please?

And one comment/question:
- counterexample printing currently prints as fractions. This is fine, but is this an artifact of how the SMT solver returns the counterexample? Does the smt solver always return a fraction?

lsk567 commented on src/main/scala/uclid/lang/UclidParser.scala:
```diff
@@ -66,6 +66,10 @@ trait UclidTokenParsers extends TokenParsers {
 
+  /** A parser which matches a real literal */
+  def realLit: Parser[RealLit] =
+    elem("real", _.isInstanceOf[RealLit]) ^^ (_.asInstanceOf[RealLit])
```
Thanks for another good catch. `RealLit` is now defined in `UclidLexical.scala`.

In the proposed implementation, `x.y` will be first tokenized as a `realLit`. If there is a floating point keyword attached to it (e.g. `x.y` + `half`), then it will be later parsed as a float literal. Otherwise, it will be parsed as a real literal.

I think this is a cleaner design than tokenizing `x.y` as a `floatLit` in the lexer.

lsk567 commented on src/main/scala/uclid/lang/UclidLexical.scala:
```diff
@@ -115,7 +115,7 @@ class UclidLexical extends Lexical with UclidTokens with Positional {
     | positioned { '0' ~ 'x' ~> rep1( hexDigit )                       ^^ { case hits => IntegerLit(hits.mkString, 16) } }
     | positioned { '0' ~ 'b' ~> rep1( bit )                            ^^ { case bits => IntegerLit(bits.mkString, 2) } }
-    | positioned { digit ~ rep(digit) ~ '.' ~ digit ~ rep(digit) ^^ { case (f1 ~ r1 ~ '.' ~ f2 ~ r2) => FloatLit(f1 :: r1 mkString, f2 :: r2 mkString,11,52) } }
+    | positioned { digit ~ rep(digit) ~ '.' ~ digit ~ rep(digit) ^^ { case (f1 ~ r1 ~ '.' ~ f2 ~ r2) => FloatLit(f1 :: r1 mkString, f2 :: r2 mkString,11,52) } }    
```
Fixed.

lsk567 commented:
> Hey Shaokai, I probably missed this, but did we discuss `r` as an alternative to the `real` suffix (`real` is kind of cumbersome)?

Thanks for the prompt review, @adwait.

In the updated implementation, real numbers just appear as `x.y` without any suffixes. The data type "real" is still `real`. I think this design looks better than using either `real` or `r` as a suffix. Here is an example:
```
module main {
  // Part 1: System description.
  var a, b : real; // Declaring a and b as real variables.

  init {
    a = 0.0; // 0.0 is a real literal.
    b = 1.1; // 1.1 is a real literal.
  }
  next {
    a', b' = b, a + b;
  }

  // Part 2: System specification.
  invariant a_le_b: a > b; // Expect SAT.

  // Part 3: Proof script.
  control {
    v = bmc(2);
    check;
    print_results;
    v.print_cex;
  }
}
```

lsk567 commented:
Thanks for the detailed review, @polgreen!

> * I'm a bit confused about where RealLit is defined. I think it needs to be in UclidLexical, so I wonder if you've just forgotten to commit a line? As it is currently written, it throws a compiler warning saying that "fruitless type test: a value of type UclidTokenParsers.this.Elem cannot also be a uclid.lang.RealLit"

This is now fixed. In the new commit, `x.y` will be first tokenized as a `realLit`. If a floating point keyword is detected during parsing (e.g. x.y + half), then it will be further parsed as a float literal. Otherwise, it will be parsed as a real literal.

> * add reals to this match in the TypeChecker: https://github.com/uclid-org/uclid/blob/reals/src/main/scala/uclid/lang/TypeChecker.scala#L222
> * and this: https://github.com/uclid-org/uclid/blob/reals/src/main/scala/uclid/smt/Context.scala#L153
> 
> Also, some more testing please!
> 
> * Can you add some tests that should fail verification as well as tests that pass verification?
> * Can you also add the tests to run using the Z3 interface as well as the SMTlib interface please?

Fixed!

> * Also here (actually this one also fails on float, so if you could add float as well that would be fantastic): https://github.com/uclid-org/uclid/blob/reals/src/main/scala/uclid/smt/Z3Interface.scala#L306

Let's fix float in another PR?

> And one comment/question:
> 
> * counterexample printing currently prints as fractions. This is fine, but is this an artifact of how the SMT solver returns the counterexample? Does the smt solver always return a fraction?

Yes, it appears that Z3 always returns a fraction.
